frontend/src/metabase/visualizations/components/Visualization/Visualization.jsx

```diff
@@ -72,39 +72,87 @@ const mapStateToProps = state => ({
 
 const SMALL_CARD_WIDTH_THRESHOLD = 150;
 
-class Visualization extends PureComponent {
-  state = {
-    getHref: undefined,
-    hovered: null,
-    clicked: null,
-    error: null,
-    genericError: null,
-    warnings: [],
-    series: null,
-    visualization: null,
-    computedSettings: {},
+const isLoading = series => {
+  return !(
+    series &&
+    series.length > 0 &&
+    _.every(
+      series,
+      s => s.data || _.isObject(s.card.visualization_settings.virtual_card),
+    )
+  );
+};
```

Comment:
Extracts `isLoading` to a separate pure function, so it can be referenced in `getDerivedStateFromProps`